plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+            subThread->wait(TIMEOUT);//locked so lets subscribe for value
+        else
+            return;//normal GET
+    }
+}
+void AsyncConnection::handleLockForSet(ICodeContext * ctx, const char * key, const char * value, size_t size, const char * channel, unsigned expire)
+{
+    StringBuffer cmd("SET %b %b");
+    RedisPlugin::appendExpire(cmd, expire);
+    attachLibev();
+    if(channel)//acknowledge locks i.e. for use as the race winner and lock owner
+    {
+        //Due to locking logic surfacing into ECL, any locking.set (such as this is) assumes that they own the lock and therefore just go ahead and set
+        //It is possible for a process/call to 'own' a lock and store this info in the LockObject, however, this prevents sharing between clients.
+        assertRedisErr(redisAsyncCommand(context, setCB, NULL, cmd.str(), key, strlen(key), value, size), "SET buffer write error");
+        handleLoop(EV_DEFAULT_ EV_TIMEOUT);//not theoretically necessary as subscribers receive value from published message. In addition,
+        //the logic stated above allows for any other client to set and therefore as soon as this is set it can be instantly altered.
+        //However, this is here as libev handles socket io to/from redis.
+        assertRedisErr(redisAsyncCommand(context, pubCB, NULL, "PUBLISH %b %b", channel, strlen(channel), value, size), "PUBLISH buffer write error");
+        handleLoop(EV_DEFAULT_ EV_TIMEOUT);//this only waits to ensure redis received pub cmd. MORE: reply contains number of subscribers on that channel - utilise this?
+    }
+    else
+    {
+        ReturnValue retVal;
+        //This branch represents a normal async get i.e. no lockObject present and no channel
+        //There are two primary options that could be taken    1) set regardless of lock (publish if it was locked)
+        //                                                     2) Only set if not locked
+        //(1)
+        StringBuffer cmd2("GETSET %b %b");
+        RedisPlugin::appendExpire(cmd2, expire);
+        //obtain channel
+        assertRedisErr(redisAsyncCommand(context, getCB, (void*)&retVal, cmd2.str(), key, strlen(key), value, size), "SET buffer write error");
+        handleLoop(EV_DEFAULT_ EV_TIMEOUT);
+        if (strncmp(retVal.str(), REDIS_LOCK_PREFIX, strlen(REDIS_LOCK_PREFIX)) == 0 )
+        {
+            assertRedisErr(redisAsyncCommand(context, pubCB, NULL, "PUBLISH %b %b", channel, strlen(channel), value, size), "PUBLISH buffer write error");
+            handleLoop(EV_DEFAULT_ EV_TIMEOUT);//again not necessary, could just call redisAsyncHandleWrite(context);
+        }
+    }
+}
+//-------------------------------------------GET-----------------------------------------
+//---OUTER---
+template<class type> void AsyncRGet(ICodeContext * ctx, const char * options, const char * key, type & returnValue, const char * channel, unsigned __int64 database)
```

Comment:
What are the async versions meant to do?  Is there any advantage to the ECL user?  Should we consider an async function which evaluates early, but is then queried to extract the results.  (Child queries are heading this way with lazy evaluation.)